### PR TITLE
Delegation

### PR DESCRIPTION
**Pull request by malakai97**
This adds support for the hosted* and proxied delegation scenarios. *The hosted system tests are currently disabled because I wasn't able to put the data into the headers.

API:
We add a collection repo. All requests begin with a call to this repo so that we can use the collection's mode to then call either delegated or normal mode-related code.

You'll notice a "hand-made" factory in test/support/fabricators. This was done to get around an apparent issue with rom-factory where associations were not being assigned correctly. You might recall I ranted about this in slack. 

There is also a change in the Gemfile to pull down my fork of rom-sql. Basically, the method that gets the response back from Sequel / the mysql client relies on that client to tell it what the primary keys were. However, it does not know unless they're auto-incrementing, numeric keys. So, it just numbers the rows it returns. Do you have one row returned? That's 0. This was causing the return value of ROM `create` calls to be...weird. My small change relies on the fact that the method in question already knows what the primary key is, and simply uses that if the client told is the pk was 0. I believe we could workaround this at the cost of more code. I wouldn't want to leave things in this state without setting up some sort of automatic sync on the fork. (It's set to a tag now.)

Docker:
Adds proxied-test-app, which is a python script listening on 8008 that simply puts the headers into the response body. This is only used in the proxied scenarios (currently disabled).

DB:
Adds delegation.sql. This constructs a number of collections related by dlpsClass for use in system tests. 

Module:
The module now sets the AUTHZD_COLL and PUBLIC_COLL environment variables. It does not manipulate these values; just assigns the strings it receives. See the client section below.

Apache config:
* Enable mod proxy and setup a proxy location at /app/proxy. 
* Setup a location for hosted (cgi) scenarios at /hosted. Uses the printenv script.

Client:
* Expands AuthorizationResult to hold the two lists of collections. 
* Removes Authorizer#isAllowed.
* Adds Authorizer#authorize. This gets and AuthorizationResult from the ApiClient, and converts it to a `map<string,string>` that has the final form of the collections, i.e. "coll1:coll2:coll3".

### CODE REVIEW

File: lauth/app/ops/authorize.rb [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+      def call(request:)
+        collection = collection_repo.find_by_uri(request.uri)
+        case collection.dlpsAuthzType
+        when "n"
+          normal_mode(request: request)
+        when "d"
+          delegated_mode(request: request, collection: collection)
+        else
+          raise "Unknown dlpsAuthzType '#{collection.dlpsAuthzType}'"
```
Comment (botimer) [inline lauth/app/ops/authorize.rb:17]: Probably want a typed error if raising is the right path

File: lauth/app/repositories/collection_repo.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
+module Lauth
+  module Repositories
+    class CollectionRepo < ROM::Repository[:collections]
+      include Deps[container: "persistence.rom"]
+      struct_namespace Lauth
+
+      def find(id)
+        grants.where(uniqueIdentifier: id).one
```
Comment (botimer) [inline lauth/app/repositories/collection_repo.rb:8]: collections -- is this unused/untested?

File: lauth/spec/repositories/grant_repo_spec.rb [hunk 3]
```diff
@@ -1,150 +1,168 @@
...
+    let!(:collection_b) { Factory[:collection] }
+    let!(:grant_a) { Factory[:grant, :for_user, user: user, collection: collection_a] }
+    let!(:grant_b) { Factory[:grant, :for_user, user: user, collection: collection_b] }
+    it "finds grants for other collections with the same dlpsClass" do
+      grants = repo.for_collection_class(
+        username: "lauth-allowed", collection_class: collection_a.dlpsClass, client_ip: "1.2.3.4")
+
+      expect(grants.map(&:uniqueIdentifier)).
+        to contain_exactly grant_a.uniqueIdentifier, grant_b.uniqueIdentifier
+    end
```
Comment (botimer) [inline lauth/spec/repositories/grant_repo_spec.rb:16]: Clarify the purpose here -- finding within the class explicitly

File: lauth/spec/requests/delegated_spec.rb [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+  before(:each) do
+    setup_coll("secret", has_grant: false, pub: false)
+    setup_coll("public", has_grant: false, pub: true)
+    setup_coll("extra",  has_grant: true,  pub: false)
+    setup_coll("both",   has_grant: true,  pub: true)
+    target = setup_coll("target", has_grant: true,  pub: false)
+    Factory[:location, dlpsPath: "/delegated", collection: target]
+  end
+
+  context "when authorized" do
```
Comment (botimer) [inline lauth/spec/requests/delegated_spec.rb:12]: what is this actual scenario? maybe "with an authorized, delegated collection" or so?

File: test-site/app/proxied.py [hunk 5]
```diff
@@ -0,0 +1,16 @@
+from http.server import HTTPServer, BaseHTTPRequestHandler
```
Comment (botimer) [inline test-site/app/proxied.py:1]: Need to include license and acknolwedgment

File: test/delegation/hosted_spec.rb [hunk 6]
```diff
@@ -1,4 +1,58 @@
...
+      expect(response.status).to eq HttpCodes::OK
+    end
+
+    it "lists the matching public collections" do
+      expect(parse_env(response.body)["PUBLIC_COLL"]&.split(":"))
+        .to contain_exactly 'public-cats'
+    end
+
+    it "lists the matching authorized collections" do
+      expect(parse_env(response.body)["AUTHZD_COLL"]&.split(":"))
```
Comment (botimer) [inline test/delegation/hosted_spec.rb:22]: Ensure we've dealt with `:`, `:one:`, `:one:two:`

**Review by botimer — CHANGES_REQUESTED** [review]

File: lauth/app/ops/authorize.rb (hunk 1 above)
Comment (malakai97) [reply lauth/app/ops/authorize.rb:17]: It should be unreachable but it isn't sufficiently constrained that I didn't want the safeguard in place for future development.

File: notes [hunk 7]
```diff
@@ -0,0 +1,41 @@
+managed flag
```
Comment (malakai97) [inline notes:1]: lolwoops

**Review by botimer — APPROVED** [review]

**Pull request merged**